### PR TITLE
[gen_l10n] Fallback feature for untranslated messages

### PR DESCRIPTION
**Pull request by shihaohong**
# Description

Uses the template ARB file's language for a message as a fallback for `gen_l10n` tool users and generates a report of the untranslated messages for a Flutter project.

## Sample use case

For example, if the template ARB file contains: 
```
{
  "@@locale": "en",

  "helloWorld": "Hello World",
  "@helloWorld": {
    "description": "The conventional newborn programmer greeting"
  },

  "hello": "Hello {world}",
  "@hello": {
    "description": "A message with a single parameter",
    "placeholders": {
      "world": {}
    }
  }
}
```

and a translated message arb file contains:
```
{
  "@@locale": "zh",
  "helloWorld": "你好世界",
}
```

Calling `AppLocalizations.of(context).hello('test')` in the app with a device's language set to Chinese would use the message structure defined in the template arb file.

## Generating report of untranslated message files

Using the `untranslated-messages-file` option, this creates a JSON file containing a list of messages that do not exist for particular locales. For example:

```
{
  "es": [
    "subtitle"
  ]
}
```

With `subtitle` being the resource ID for the locale that is missing a translation.

If `untranslated-messages-file` is not specified but untranslated messages exist in the app, a summary is printed in the command line:
```
"es": 1 untranslated message(s)
To see a detailed report, use the untranslated-messages-file 
option in the tool to generate a JSON format file containing 
all messages that need to be translated.
```

## Related Issues

Addresses part of https://github.com/flutter/flutter/issues/51778

## Tests

I added the following tests:
- A unit test to ensure that the txt file is created and contains the locale's name, as well as the resource id for the missing translation.
- The extra arb messages for `app_zh.arb` in the integration tests should no longer be necessary, since it uses the fallback messages
- An integration test to ensure that a fallback message is correctly used

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by guidezpl** [comment]
What did you think about embedding these warnings as console warnings? Either in the base class, or in the locale subclasses (to enable locale-specific errors). Something like "Message x has not been translated for locale Y. Using locale <default> instead". I'm just worried from a DX perspective about putting errors in a file that's sort of hidden

**Comment by shihaohong** [comment]
@guidezpl I thought about that, and my main concern is that with a ton of messages, the console warnings can be extremely verbose. Imagine having hundreds of untranslated messages and having 100s of lines of the console saying different messages do not exist.

However, it does seem like it might be valuable to maybe do a one-liner per locale? Something like "x messages for y locale have not been translated yet." And then at the end, maybe a pointer to the file like "See {fileName}.txt for the entire list of messages that require translations."

**Comment by guidezpl** [comment]
@shihaohong Fair point. I think that's a good idea. For the output file, it makes more sense to locate it in the top-level `build` directory to avoid being checked in.

**Review by HansMuller — COMMENTED** [review]
This will be a useful feature since many projects will probably incorporate translations for different locales at different rates. Some thoughts about the design:

 - Unless I'm mistaken, this PR introduces a fallback feature: if a translation doesn't exist, then we fallback to the template bundle. The PR's description should cover that.
 - I think we want to distinguish between unimplemented messages and messages that exist in the template arb file but are missing translations for some locales. For the former (per https://github.com/flutter/flutter/issues/51778#issuecomment-593538062) we want developers to be able to say AppLocalizations.foo or AppLocalizations.bar("arg1", "arg2"), before foo or bar have been defined in the template arb file. If you're planning to handle that feature in a separate PR, please say as much in the description.
 - The tool should include a parameter for specifying where the unimplemented translation summary gets written. Since tools may consume this file, you'll probably want to make sure that the format is easy to parse and very consistent.
 - If the user doesn't specify an output file, then we should probably write to the console. Doing that will require some care, since - as you pointed out - the output might be voluminous. It might be best to begin the console output with a message that reminds users that there are unimplemented translations (summarize how many messages in how many locales), that untranslated messages will fallback to the template file's translation, and that a complete report can be generated by specifying `new command line arugment here`. You could follow that with a brief summary of the the locales with missing translations.

**Comment by shihaohong** [comment]
> - Unless I'm mistaken, this PR introduces a fallback feature: if a translation doesn't exist, then we fallback to the template bundle. The PR's description should cover that.

Done.

> - I think we want to distinguish between unimplemented messages and messages that exist in the template arb file but are missing translations for some locales. For the former (per https://github.com/flutter/flutter/issues/51778#issuecomment-593538062) we want developers to be able to say AppLocalizations.foo or AppLocalizations.bar("arg1", "arg2"), before foo or bar have been defined in the template arb file. If you're planning to handle that feature in a separate PR, please say as much in the description.

Sounds good, I hadn't considered the possibility of implementing both separately, and had thought of both solutions as different ways to solve a singular problem. I've created [a separate issue](https://github.com/flutter/flutter/issues/53767) to track that. 

> - The tool should include a parameter for specifying where the unimplemented translation summary gets written. Since tools may consume this file, you'll probably want to make sure that the format is easy to parse and very consistent.

@HansMuller are there any suggestions for how to format this? I basically just did something simple:
```
Locales with missing message translations:

Locale ({LocaleCode}):
#1: {message1}
#2: {message2}
....
#n: {messageN}

Locale ({LocaleCode}):
#1: {message1}
#2: {message2}
....
#n: {messageN}
```

Are there specific ways or guidelines for formatting files for tool consumption? Maybe a JSON format? Something like:
```
{
  "LocaleCode1": [
    "message1",
    "message2",
  ],
  "LocaleCode2": [
    "message1",
    "message2",
  ],
}
```

> - If the user doesn't specify an output file, then we should probably write to the console. Doing that will require some care, since - as you pointed out - the output might be voluminous. It might be best to begin the console output with a message that reminds users that there are unimplemented translations (summarize how many messages in how many locales), that untranslated messages will fallback to the template file's translation, and that a complete report can be generated by specifying `new command line argument here`. You could follow that with a brief summary of the the locales with missing translations.

re: output file and displaying on the console, I can make this into an option for the executable. Should this mean that the default should be writing a summarized version of the output to the console? @guidezpl, who might have some thoughts on this as well, since he recommended putting the file in the `build/` folder of the project directory by default.

**Comment by HansMuller** [comment]
@shihaohong - I was suggesting making the name of the (optional) report output file an executable command-line flag. I think JSON is human-readable enough

**Comment by guidezpl** [comment]
@shihaohong Summary to console + optional json report as a command-line flag for the executable will work well for us

File: dev/tools/localization/bin/gen_l10n.dart [hunk 1]
```diff
@@ -40,6 +40,15 @@ void main(List<String> arguments) {
     help: 'The filename for the output localization and localizations '
       'delegate classes.',
   );
+  parser.addOption(
+    'untranslated-messages-file',
+    help: 'The location of generate the file describing which localization\n'
```
Comment (HansMuller) [inline dev/tools/localization/bin/gen_l10n.dart:45]: the location of a file that describes the localization

File: dev/tools/localization/gen_l10n.dart [hunk 2]
```diff
@@ -741,4 +769,52 @@ class LocalizationsGenerator {
...
     outputFile.writeAsStringSync(generateCode());
   }
+
+  void outputUnimplementedMessages(String untranslatedMessagesFile) {
+    if (untranslatedMessagesFile == null || untranslatedMessagesFile == '') {
+      _unimplementedMessages.forEach((LocaleInfo locale, List<String> messages) {
+        stdout.writeln('"$locale": ${messages.length} untranslated message(s).');
+      });
+      stdout.writeln(
+        'To see a detailed report, use the unimplemented_messages_file \n'
```
Comment (HansMuller) [inline dev/tools/localization/gen_l10n.dart:779]: --unimplemented-messages-file

**Review by HansMuller — APPROVED** [review]
LGTM modulo a few typos

**Comment by shihaohong** [comment]
I made some updates to the PR because I realized that testing could be further improved and the PR description could be more descriptive. @HansMuller PTAL

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 3]
```diff
@@ -129,6 +129,9 @@ class Home extends StatelessWidget {
             results.add(AppLocalizations.of(context).helloWorlds(0));
             results.add(AppLocalizations.of(context).helloWorlds(1));
             results.add(AppLocalizations.of(context).helloWorlds(2));
+            // Should use the fallback language, in this case,
+            // "Hello 世界" should be displayed.
+            results.add(AppLocalizations.of(context).hello("世界"));
```
Comment (HansMuller) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart:134]: NICE

**Review by HansMuller — APPROVED** [review]

**Pull request merged**